_posts/2024-11-12-a-kernel-story7.md

```diff
@@ -0,0 +1,156 @@
+
+<div style="text-align: center;" markdown="1">
+| ![Github Actions](/assets/figures/2024-11-12-Setting_CS_to_active_high_1.jpg) | 
+|:--:| 
+| *Generated by Gemini AI* |
+</div>
+
+# Changing the device-tree (DT) overlay
+
+The necessary configuration to operate this change occurs in the DT, and this is what I am going to explain in this section.
+
+## Explanation
+
+To invert the CS signal we need to set the `cs-gpio` to `ACTIVE_HIGH` in the controller **AND** set the `spi-cs-high` in the peripheral property. The kernel documentation defines a table where this is explained<sup>[1](#references)</sup>. Let me quote it:
+
+```
+There is a special rule set for combining the second flag of an cs-gpio with the optional spi-cs-high flag for SPI slaves.
+
+Each table entry defines how the CS pin is to be physically driven (not considering potential gpio inversions by pinmux):
+
+device node     | cs-gpio       | CS pin state active | Note
+================+===============+=====================+=====
+spi-cs-high     | -             | H                   |
+-               | -             | L                   |
+spi-cs-high     | ACTIVE_HIGH   | H                   |
+-               | ACTIVE_HIGH   | L                   | 1
+spi-cs-high     | ACTIVE_LOW    | H                   | 2
+-               | ACTIVE_LOW    | L                   |
+```
+
+We are interested in the third line of the table since it is the one that sets the CS pin to `ACTIVE_HIGH`. Therefore, and as mentioned above, we need to set both the `cs-gpio` to `ACTIVE_HIGH` and set the `spi-cs-high` property.
+
+## Hands-on
+
+In the Fedora distribution, overlays are located in the `/boot/efi/overlays` folder. The files in this folder are in a format understood by the Linux kernel, so you need to convert them to a human-readable format first. As an example, you can convert `spi0-1cs` with the following command:
+
+```console
+dtc -I dtb -O dts spi0-1cs.dtbo -o spi0-1cs.dtbs
+```
+
+However, I wouldn't recommend you to do this transformation, but to download the file from the Raspberry Pi repositories<sup>[2](#references)</sup>, because the transformation loses some details and the final result could be wrong.
+
+Now let's focus on the changes. First, you need to set the `cs-gpios` to `ACTIVE_HIGH` (or 0) in the `fragment@1` `__overlay__`. Example:
+
+```
+fragment@1 {
+    target = <&spi0>;
+    frag1: __overlay__ {
+        cs-gpios = <&gpio 8 0>;
+        status = "okay";
+    };
+};
+```
+
+Second, add a new `fragment` whose target is the user-space `<&spidev0>` peripheral with the `spi-cs-high` property. Example:

```

Comment:
instead of the user-space maybe you can say "the virtual spidev0 device, these devices are exposed to user-space through a `/dev/spidevX` device node to allow user-space drivers to access the SPI peripherals directly. This is useful when there are no Linux kernel drivers for the devices". Or something along those lines.

Reply:
Good one. Thanks!